p2p/net/connmgr/connmgr.go

```diff
@@ -246,15 +253,32 @@ type peerInfo struct {
 	firstSeen time.Time // timestamp when we began tracking this peer.
 }
 
-type peerInfos []peerInfo
+type peerInfos []*peerInfo
 
 // SortByValueAndStreams sorts peerInfos by their value and stream count. It
 // will sort peers with no streams before those with streams (all else being
 // equal). If `sortByMoreStreams` is true it will sort peers with more streams
 // before those with fewer streams. This is useful to prioritize freeing memory.
-func (p peerInfos) SortByValueAndStreams(sortByMoreStreams bool) {
+func (p peerInfos) SortByValueAndStreams(segments *segments, sortByMoreStreams bool) {
 	sort.Slice(p, func(i, j int) bool {
 		left, right := p[i], p[j]
+
+		// Grab this lock so that we can grab both segment locks below without deadlocking.
+		segments.bucketsMu.Lock()
+
+		// lock this to protect from concurrent modifications from connect/disconnect events
+		leftSegment := segments.get(left.id)
+		leftSegment.Lock()
```

Comment:
I know this might be very tricky. But I not very sure about the fine grained locking, does locking at bucketMu level make more sense?

Reply:
1. We need to get the segment locks because something else (like a connection notification event) can be writing to the conns map.
2. We may need to get 2 segment locks, so the `bucketMu` protects us from a deadlock when grabbing the 2 locks.

---

If we only locked the bucketMu here we wouldn't fix anything since writers can still modify the conns map. If we make writers also take the bucketMu, then there's no point in the segmented locks anymore and possibly regress in performance here: https://github.com/libp2p/go-libp2p-connmgr/pull/40.

We _may_ no longer need the segmented locks, but without some benchmarks I would be hesitant to change this.

Reply:
Thanks for the pointer to the segment lock history. That make sense now. 

I did a quick search on the code, did not find other cases of needing two segment locks at the same time (other than the sort here). Hope that is an exhaustive search and the condition holds here.  Anyway we may need to come back and review this for a better option such as CSP.